core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -333,6 +346,181 @@ public long nextFetchOffset() {
+                        if (!fullMatch || inFlightBatch.offsetState != null) {
+                            log.debug("Subset or offset tracked batch record found while trying to update offsets and cached" +
+                                            " state map due to LSO movement, batch: {}, offsets to update - " +
+                                            "first: {}, last: {} for the share partition: {}-{}", inFlightBatch, startOffset,
+                                    logStartOffset - 1, groupId, topicIdPartition);
+
+                            if (inFlightBatch.offsetState == null) {
+                                if (inFlightBatch.batchState() == RecordState.ARCHIVED) {
+                                    continue;
+                                }
+                                inFlightBatch.maybeInitializeOffsetStateUpdate();
+                            }
+                            throwable = archivePerOffsetBatchRecords(inFlightBatch, startOffset, logStartOffset - 1,
+                                    updatedStates, stateBatches).orElse(null);
+                            if (throwable != null) {
+                                break;
+                            }
+                        } else {
+                            // The in-flight batch is a full match hence change the state of the complete batch.
+                            throwable = archiveCompleteBatch(inFlightBatch, updatedStates, stateBatches).orElse(null);
+                            if (throwable != null) {
+                                break;
+                            }
+                        }
+                    }
+                    // If the transitions to archive is successful then persist state, complete the state transition
+                    // and update the cached state for start offset. Else rollback the state transition.
+                    rollbackOrProcessArchivedStates(throwable, updatedStates, stateBatches);
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+        }
+    }
+
+
+    // This function provides the logic to get the earliest valid offset for a topic partition.
+    private long offsetForEarliestTimestamp() {
+        // TODO: We need to know the isolation level from group configs, for now we are passing Option.empty() for isolationLevel
+        Option<FileRecords.TimestampAndOffset> timestampAndOffset = replicaManager.fetchOffsetForTimestamp(
+                topicIdPartition.topicPartition(), ListOffsetsRequest.EARLIEST_TIMESTAMP, Option.empty(),
+                Optional.empty(), true);
+        return timestampAndOffset.isEmpty() ? (long) 0 : timestampAndOffset.get().offset;
+    }
+
+    private Optional<Throwable> archivePerOffsetBatchRecords(InFlightBatch inFlightBatch,
+                                                             long startOffset,
```

Comment:
These parameters `startOffset` and `endOffset` have the same names as the member variables in this class. That's not a very good idea in terms of code readability I think.